### PR TITLE
add filter visual

### PR DESCRIPTION
**Pull request by Gustav-Eikaas**
# Description

Please include a summary of the change and which issue is fixed.
List any external dependencies that are required for this change.

Completes: AB#FILL_IN_YOUR_ISSUE_ID

## Type of change

Remove options that are not relevant.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my own code.
-   [ ] I have linked my DevOps task using the AB# tag.
-   [x] My code is easy to read.
-   [x] Documentation and comments is added where needed.
-   [x] My code is covered by tests.

### CODE REVIEW

File: packages/filter/src/lib/classes/countCache.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+export class CountCache {
```
Comment (Gustav-Eikaas) [inline packages/filter/src/lib/classes/countCache.ts:1]: docs

Comment (Gustav-Eikaas) [reply packages/filter/src/lib/classes/countCache.ts:1]: Why do we have countcache

File: packages/filter/src/lib/classes/reactFilterController.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import { FilterConfiguration, ValueFormatterFilter } from '../types';
+import { FilterController } from './filterController';
+import { Observable, OnchangeCallback } from './observable';
+
+export class ReactFilterController<TData> extends FilterController<TData> {
```
Comment (Noggling) [inline packages/filter/src/lib/classes/reactFilterController.ts:5]: doc

File: packages/filter/src/lib/components/ExpandedFilterGroup/ExpandedFilterGroup.tsx [hunk 3]
```diff
@@ -0,0 +1,163 @@
+import { Icon, Search } from '@equinor/eds-core-react';
```
Comment (Gustav-Eikaas) [inline packages/filter/src/lib/components/ExpandedFilterGroup/ExpandedFilterGroup.tsx:1]: Cleanup

File: packages/filter/src/lib/components/ExpandedFilterGroup/ExpandedFilterGroup.tsx [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+		</Wrapper>
+	);
+};
+
+interface VirtualContainerProps {
+	filterGroup: FilterGroup;
+	filterSearchValue: string;
+}
+
+export const VirtualContainer = ({ filterGroup, filterSearchValue }: VirtualContainerProps): JSX.Element | null => {
```
Comment (Noggling) [inline packages/filter/src/lib/components/ExpandedFilterGroup/ExpandedFilterGroup.tsx:114]: Splitt to own file

File: packages/filter/src/lib/components/ExpandedFilterGroup/expandedFilterGroup.styles.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+		`2px solid ${
+			isActive ? tokens.colors.interactive.primary__resting.hex : tokens.colors.ui.background__medium.hex
+		}`};
+
+	color: ${({ isActive }) =>
+		isActive ? tokens.colors.interactive.primary__resting.hex : tokens.colors.text.static_icons__default.hex};
+
+	margin-bottom: 5px;
+`;
+
```
Comment (Gustav-Eikaas) [inline packages/filter/src/lib/components/ExpandedFilterGroup/expandedFilterGroup.styles.ts:89]: prefix  styled

File: packages/filter/src/lib/components/Filter.tsx [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import { useFilterContext } from '../hooks';
+import { QuickFilter } from './quickFilter/QuickFilter';
+import { playlist_add, search, drag_handle } from '@equinor/eds-icons';
+import { Icon } from '@equinor/eds-core-react';
+Icon.add({ playlist_add, search, drag_handle });
+
+export function Filter() {
```
Comment (Noggling) [inline packages/filter/src/lib/components/Filter.tsx:7]: docs

File: packages/filter/src/lib/components/SearchPickerDropdown.tsx [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+interface MenuItem {
+	title: string;
+	onCLick: () => void;
+}
+interface SearchPickerDropdownProps {
+	menuItems: MenuItem[];
+}
+
+export function SearchPickerDropdown({ menuItems }: SearchPickerDropdownProps): JSX.Element {
```
Comment (Noggling) [inline packages/filter/src/lib/components/SearchPickerDropdown.tsx:14]: Docs

File: packages/filter/src/lib/components/ToggleHideFilterPopover.tsx [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+};
+
+const PopoverList = styled.div`
+	display: flex;
+	flex-direction: column;
+	align-items: center;
+	padding-bottom: 20px;
+`;
+
+const ItemWrapper = styled.div`
```
Comment (Noggling) [inline packages/filter/src/lib/components/ToggleHideFilterPopover.tsx:90]: prefix styled

File: packages/filter/src/lib/components/ToggleHideFilterPopover.tsx [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+							</ReactSortable>
+						</PopoverList>
+					</Popover.Content>
+				</Popover>
+			)}
+		</>
+	);
+};
+
+const PopoverList = styled.div`
```
Comment (Gustav-Eikaas) [inline packages/filter/src/lib/components/ToggleHideFilterPopover.tsx:83]: prefix and move to styles file

File: packages/filter/src/lib/components/ToggleHideFilterPopover.tsx [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+import styled from 'styled-components';
+
+interface ShowHideFilterButtonProps {
+	allFilters: string[];
+	visibleFilters: string[];
+	setVisibleFilters: (val: string[]) => void;
+}
+
+interface SortObject<T> {
+	id: string;
```
Comment (Gustav-Eikaas) [inline packages/filter/src/lib/components/ToggleHideFilterPopover.tsx:14]: move

File: packages/filter/src/lib/components/ToggleHideFilterPopover.tsx [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+import { ReactSortable } from 'react-sortablejs';
+import styled from 'styled-components';
+
+interface ShowHideFilterButtonProps {
+	allFilters: string[];
+	visibleFilters: string[];
+	setVisibleFilters: (val: string[]) => void;
+}
+
+interface SortObject<T> {
```
Comment (Noggling) [inline packages/filter/src/lib/components/ToggleHideFilterPopover.tsx:13]: move

File: packages/filter/src/lib/components/filterGroup/FilterGroupPopoverMenu.tsx [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+
+	const handleInput = (e) => setSearchText(e.target.value.toString().toLowerCase());
+
+	const getValuesMatchingSearchText = () =>
+		values.filter((s) => !searchText || s?.toString().toLowerCase().startsWith(searchText));
+
+	const setFilterStateFromSearch = () => {
+		const valuesMatchingSearch = values.filter((s) => !getValuesMatchingSearchText().includes(s));
+
+		setFilterState([
```
Comment (Noggling) [inline packages/filter/src/lib/components/filterGroup/FilterGroupPopoverMenu.tsx:45]: doc

File: packages/filter/src/lib/components/filterGroup/FilterGroupPopoverMenu.tsx [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+	const { valueFormatters, getCountForFilterValue } = useFilterContext();
+	const { setFilterState, filterState } = useFilterState();
+	const valueFormatter = valueFormatters.find(({ name }) => name === groupName)?.valueFormatter;
+
+	const handleInput = (e) => setSearchText(e.target.value.toString().toLowerCase());
+
+	const getValuesMatchingSearchText = () =>
+		values.filter((s) => !searchText || s?.toString().toLowerCase().startsWith(searchText));
+
+	const setFilterStateFromSearch = () => {
```
Comment (Gustav-Eikaas) [inline packages/filter/src/lib/components/filterGroup/FilterGroupPopoverMenu.tsx:42]: explain

File: packages/filter/src/lib/components/filterGroup/FilterGroupPopoverMenu.tsx [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+									}
+								}}
+							/>
+						</SearchHolder>
+						<VerticalLine />
+					</>
+				)}
+
+				<List>
+					{getValuesMatchingSearchText().map((value) => (
```
Comment (Noggling) [inline packages/filter/src/lib/components/filterGroup/FilterGroupPopoverMenu.tsx:85]: virtualize?

File: packages/filter/src/lib/components/filterGroup/FilterGroupPopoverMenu.tsx [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+					<Button onClick={markAllValuesActive} variant="ghost">
+						Clear
+					</Button>
+				</ClearButtonWrapper>
+			</MenuWrapper>
+		</Menu>
+	);
+};
+
+const List = styled.div`
```
Comment (Gustav-Eikaas) [inline packages/filter/src/lib/components/filterGroup/FilterGroupPopoverMenu.tsx:108]: move to file, prefix

File: packages/filter/src/lib/components/filterGroup/FilterGroup.tsx [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+					values={values}
+					CustomRender={customRender}
+					groupName={name}
+				/>
+			)}
+		</div>
+	);
+};
+
+export function getFilterHeaderText(
```
Comment (Noggling) [inline packages/filter/src/lib/components/filterGroup/FilterGroup.tsx:75]: move

File: packages/filter/src/lib/components/filterGroup/filterGroup.styles.ts [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+	&:hover {
+		background-color: ${tokens.colors.interactive.primary__hover_alt.hex};
+	}
+`;
+
+export const NormalText = styled.div`
+	font-weight: 400;
+`;
+
+export const VerticalLine = styled.div`
```
Comment (Noggling) [inline packages/filter/src/lib/components/filterGroup/filterGroup.styles.ts:29]: Prefix all !

File: packages/filter/src/lib/components/filterItem/filterItem.styles.ts [hunk 18]
```diff
@@ -0,0 +1,45 @@
+import { Checkbox } from '@equinor/eds-core-react';
+import { tokens } from '@equinor/eds-tokens';
+import styled from 'styled-components';
+
+export const AllCheckbox = styled(Checkbox)`
```
Comment (Noggling) [inline packages/filter/src/lib/components/filterItem/filterItem.styles.ts:5]: prefix all!

File: packages/filter/src/lib/components/filterView/filterView.styles.ts [hunk 19]
```diff
@@ -0,0 +1,96 @@
+import { Button } from '@equinor/eds-core-react';
+import { tokens } from '@equinor/eds-tokens';
+import styled from 'styled-components';
+
+export const Wrapper = styled.section`
```
Comment (Noggling) [inline packages/filter/src/lib/components/filterView/filterView.styles.ts:5]: Prefix all!

File: packages/filter/src/lib/components/quickFilter/QuickFilter.tsx [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+/**
+ * How to separate controller and visual logic in this component?
+ */
+interface QuickFilterProps<T> {
+	controller: FilterController<T>;
+	groups: Omit<FilterConfiguration<T>, 'valueFormatter'>[];
+}
+
+export function QuickFilter<T>({ controller = new FilterController(), groups }: QuickFilterProps<T>): JSX.Element {
+	useOnFilteredDataChanged();
```
Comment (Gustav-Eikaas) [inline packages/filter/src/lib/components/quickFilter/QuickFilter.tsx:27]: might be unnecessary

File: packages/filter/src/lib/components/quickFilter/QuickFilter.tsx [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+	width: 100%;
+	overflow: scroll;
+`;
+
+interface OtherFiltersAppliedInfoProps {
+	activeFilters: number;
+}
+
+export function OtherFiltersAppliedInfo({ activeFilters }: OtherFiltersAppliedInfoProps): JSX.Element | null {
+	if (activeFilters <= 0) return null;
```
Comment (Gustav-Eikaas) [inline packages/filter/src/lib/components/quickFilter/QuickFilter.tsx:123]: export, refactor

File: packages/filter/src/lib/components/quickFilter/QuickFilter.tsx [hunk 22]
```diff
@@ -0,0 +1,133 @@
...
+	useOnFilteredDataChanged();
+	const { setIsFilterExpanded } = useFilterContext();
+
+	const [filterGroupOpen, setFilterGroupOpen] = useState<string | null>(null);
+	const {
+		valueFormatters,
+		filterStateController: { filterState, getInactiveGroupValues, clearActiveFilters },
+	} = controller;
+
+	function checkHasActiveFilters() {
```
Comment (Gustav-Eikaas) [inline packages/filter/src/lib/components/quickFilter/QuickFilter.tsx:36]: extract

File: packages/filter/src/lib/components/quickFilter/quickFilterStyles.ts [hunk 23]
```diff
@@ -0,0 +1,34 @@
+import { tokens } from '@equinor/eds-tokens';
+import styled from 'styled-components';
+
+export const CompactFilterWrapper = styled.div`
```
Comment (Noggling) [inline packages/filter/src/lib/components/quickFilter/quickFilterStyles.ts:4]: prefix

File: packages/filter/src/lib/components/filterGroup/FilterGroupPopoverMenu.tsx (hunk 14 above)
Comment (Gustav-Eikaas) [reply packages/filter/src/lib/components/filterGroup/FilterGroupPopoverMenu.tsx]: We can't remember why we post-poned this. I suggest we wait until performance issues is encountered

**Review by Noggling — APPROVED** [review]
LGTM

**Pull request merged**